### DISCUSSION TITLE
Ability to edit and change Master Slide?

### DISCUSSION THREAD

**Discussion by sanjevirau** [ShapeCrawler/ShapeCrawler#208] [Q&A]
Hi there,
Thank you so much for this library. I'm having a use case where:

1. There are multiple Master Slides in my PPTX and I want to move all my slides into only One Master Slide and remove the other Master Slides.
2. I want to change the Theme Color of my PPTX programmatically. Since the first use-case makes all the Master Slide into one, thus only need to change the theme color for the one Master Slide it has.

I hope the use case is understandable. (If not, do let me know). Can ShapeCrawler help me achieve this? 

And if not, can I know if this use-case is achievable programmatically or not.

Thank you :)

**Comment by AdamShakhabov** [comment]
Hello. Thank you for feedback😊

Your requirements are clear, and, yes, that can be achieved programmatically. Unfortunately, ShapeCrawler does not have yet so rich APIs for manipulating Slide Master (Master) and Slide Layout (Layout) slides. Currently, the most ShapeCrawler's APIs support _reading_ of properties of that presentation layers.

As you might know, you should deep manipulate not only Master, but also Layout. It is being far more complicated if you are not going to just prepare template for future presentations, you edit existing production presentation with real user slides (Slide). In this case, for example, before moving Master shape to a new place, you should check whether it referenced (placeholders) by Layout or Slide. All such references should be re-mapped.

**Comment by sanjevirau** [comment]
Hi @adamshakhabov,
Thank you so much for your detailed reply.

I've somehow more/less managed to achieve what I wanted to do. My process:

1. I've created a Base pptx file with all the Slide Layouts for the slides needed with the theme color.
2. Used https://github.com/EricWhiteDev/Open-Xml-PowerTools to merge multiple pptx with the Base pptx's Master Slide.

As result, I've got a PowerPoint file with only 1 Master Slide with the corrected merged slides. However, there is only 1 issue left. There are many unused/unlinked Slide Layouts in the pptx file.

Is there any way you might know how to remove unused/unlinked Slide Layouts? This is to reduce the size of the pptx file.

Thank you so much once again! 🤗

**Comment by AdamShakhabov** [comment]
No problem!

Here is a code sample where I try to delete all layouts except one used by the first slide. Similar should be for you:

```C#
var presentation = PresentationDocument.Open(@"remove-extra-layouts.pptx", true);
var usedSlidePart = presentation.PresentationPart.SlideParts.First();
var usedSlideMaster = usedSlidePart.SlideLayoutPart.SlideMasterPart;

foreach (var layoutPart in usedSlideMaster.SlideLayoutParts.ToList())
{
    if (usedSlidePart.SlideLayoutPart != layoutPart)
    {
        usedSlideMaster.DeletePart(layoutPart);
    }
}

var sldLayoutIdList = usedSlideMaster.SlideMaster.SlideLayoutIdList;
// Remove references unused Layouts from sldLayoutIdList
// ...

presentation.Close();
```

The code is not completed. We also should delete some links from Slide Master (`sldLayoutIdList `):

![image](https://user-images.githubusercontent.com/5969304/149480107-df463d66-02d5-4ebe-8d40-b843e237e968.png)
---

Later I will try to finish this sample and will update code...

**Comment by sanjevirau** [comment]
Hi @adamshakhabov ,
Thank you so much for the help! I'll try using the code you mentioned above to clear all the unlinked layouts. 🤗

**Comment by AdamShakhabov** [comment]
Done:

```C#
using var presentation = PresentationDocument.Open(@"remove-extra-layouts.pptx", true);
var usedSlidePart = presentation.PresentationPart!.SlideParts.First();
var usedSlideMaster = usedSlidePart.SlideLayoutPart!.SlideMasterPart;
var layoutIdList = usedSlideMaster!.SlideMaster!.SlideLayoutIdList!.OfType<SlideLayoutId>();

foreach (var layoutPart in usedSlideMaster.SlideLayoutParts)
{
    if (usedSlidePart.SlideLayoutPart == layoutPart)
    {
        continue;
    }
    
    var removingRelIdValue = usedSlideMaster.GetIdOfPart(layoutPart);
    var removingLayoutId = layoutIdList.First(layoutId => layoutId.RelationshipId == removingRelIdValue);
    removingLayoutId.Remove();
    usedSlideMaster.DeletePart(layoutPart);
}
```

**Comment by sanjevirau** [comment]
Hi @adamshakhabov,
Just to give you an update. I managed to achieve what I wanted by referencing your code on top and learning extensively on the OpenXML library by Microsoft. Thanks a lot, man! Much appreciated 🤗
